Scala language support

miteshaghera opened the pull request:
Initial support for Scala.

Fix for #1413

gsmet commented:
Hi @miteshaghera ,

Thanks for this. Looks like we have some test failures:
```

miteshaghera commented:
> Hi @miteshaghera ,
> 
> Thanks for this. Looks like we have some test failures:
> 
> ```
> [INFO] Quarkus - Maven Plugin ............................. FAILURE [ 28.296 s]
> ```

I fixed for `Quarkus - Maven Plugin` but now it's failing for `Quarkus - Gradle Plugin`. By looking into [logs](url) I found following exception:

    Caused by: org.gradle.internal.resolve.ModuleVersionNotFoundException: Could not find io.quarkus:quarkus-bootstrap-parent:999-SNAPSHOT.
    Searched in the following locations:
      - file:/D:/a/1/s/devtools/gradle/target/dependencies/io/quarkus/quarkus-bootstrap-parent/999-SNAPSHOT/maven-metadata.xml
      - file:/D:/a/1/s/devtools/gradle/target/dependencies/io/quarkus/quarkus-bootstrap-parent/999-SNAPSHOT/quarkus-bootstrap-parent-999-SNAPSHOT.pom
      - file:/D:/a/1/s/devtools/gradle/target/dependencies/io/quarkus/quarkus-bootstrap-parent/999-SNAPSHOT/quarkus-bootstrap-parent-999-SNAPSHOT.jar
      - https://repo.maven.apache.org/maven2/io/quarkus/quarkus-bootstrap-parent/999-SNAPSHOT/maven-metadata.xml
      - https://repo.maven.apache.org/maven2/io/quarkus/quarkus-bootstrap-parent/999-SNAPSHOT/quarkus-bootstrap-parent-999-SNAPSHOT.pom
      - https://repo.maven.apache.org/maven2/io/quarkus/quarkus-bootstrap-parent/999-SNAPSHOT/quarkus-bootstrap-parent-999-SNAPSHOT.jar

Do I need to change anything to support scala?

gsmet commented on devtools/common/src/main/filtered/extensions.json:
```diff
@@ -601,5 +601,14 @@
+    ],
+    "groupId": "io.quarkus",
+    "artifactId": "quarkus-scala",
+    "guide": "https://quarkus.io/guides/scala"
```
Let's remove this last entry as we don't have a guide yet.

miteshaghera replied:
Removed the entry and squashed them.

gsmet reviewed (changes requested):
@miteshaghera looks good to me except for a small comment.

Could you fix it? If you can squash everything into one commit, that would be appreciated too.

In the meantime, I'll try to create a Scala project and see how it goes.

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java:
```diff
@@ -60,6 +60,7 @@
     public static final String UNDERTOW_WEBSOCKETS = "undertow-websockets";
     public static final String VERTX = "vertx";
     public static final String VERTX_WEB = "vertx-web";
+    public static final String SCALA = "scala";
```
Ah and another minor additional comment: please respect the alphabetical order.

miteshaghera replied:
Fixed it

Commit:
Initial support for scala

gsmet commented on devtools/common/src/main/java/io/quarkus/generators/SourceType.java:
```diff
@@ -68,4 +73,8 @@ private String computeTemplateFile(String genericTemplate, String templateName)
     }
+
+    public static Optional<SourceType> parse(String extension) {
+        return Arrays.stream(values()).filter(e -> e.extension.contains(extension)).findAny();
```
This won't work. The extension we're talking about is the file extension so in the case of Kotlin, it's `.kt` so it doesn't contain `kotlin`.

I think I would have a method in the enum matching the extension list itself so that we can take a decision based on the list of extensions (it won't be useful for now but we could imagine it could).

miteshaghera replied:
I have fixed it by comparing it with name.

gsmet commented on devtools/common/src/main/resources/templates/basic-rest/scala/pom-template.ftl:
```diff
@@ -0,0 +1,171 @@
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>${project_groupId}</groupId>
+    <artifactId>${project_artifactId}</artifactId>
+    <version>${project_version}</version>
```
AFAICS, this file should be `pom.xml-template.ftl`.

Otherwise, when doing:
```
mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=rest-scala \
    -DclassName="org.acme.rest.GreetingResource" \
    -Dpath="/greeting" \
    -Dextensions="scala,resteasy-jsonb"
```

I have the following error:

> [ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create (default-cli) on project standalone-pom: Template resource is missing: /templates/basic-rest/scala/pom.xml-template.ftl -> [Help 1]

I wonder if the kotlin template is currently broken by the way as you copied the structure from the Kotlin one. We should check that, once you have fixed the extension/template detection.

miteshaghera replied:
I have renamed the file for Kotlin as well.

gsmet reviewed (changes requested):
OK I tried to create a project and found a couple more issues.

Could you check what's going on?

Thanks!

miteshaghera commented:
> @miteshaghera looks good to me except for a small comment.
> 
> Could you fix it? If you can squash everything into one commit, that would be appreciated too.
> 
> In the meantime, I'll try to create a Scala project and see how it goes.

I have to add entry in `extensions.json` to make build successful for `quarkus:create` . However I omitted `guide` key as we don't have guide yet.

gsmet commented on devtools/common/src/main/filtered/extensions.json:
```diff
@@ -230,6 +230,14 @@
+    ],
+    "groupId": "io.quarkus",
+    "artifactId": "quarkus-scala"
+  },
```
Please let's respect the alphabetical order.

miteshaghera replied:
Fixed it

gsmet commented on devtools/common/src/main/java/io/quarkus/generators/SourceType.java:
```diff
@@ -68,4 +73,8 @@ private String computeTemplateFile(String genericTemplate, String templateName)
     }
+
+    public static Optional<SourceType> parse(String extension) {
+        return Arrays.stream(values()).filter(e -> extension.contains(e.name().toLowerCase())).findAny();
```
I really would prefer us having a match method for each `SourceType`. Using the enum name is really fragile as we could change it without thinking it would break the behavior.

miteshaghera replied:
I have add an another field `name` in `SourceType` to maintain name of the extension.

gsmet reviewed (changes requested):
I added a couple more comments.

Could you also add a test of project creation? Considering it was broken, we really need one. See my comment here: https://github.com/quarkusio/quarkus/pull/3542#issuecomment-521933968 .

Thanks!

miteshaghera commented:
> I added a couple more comments.
> 
> Could you also add a test of project creation? Considering it was broken, we really need one. See my comment here: [#3542 (comment)](https://github.com/quarkusio/quarkus/pull/3542#issuecomment-521933968) .
> 
> Thanks!

I have added a test project creation test in`CreateProjectMojoIT` for scala.

machi1990 commented on devtools/maven/src/test/java/io/quarkus/maven/it/DevMojoIT.java:
```diff
@@ -189,6 +189,34 @@ public void testThatTheApplicationIsReloadedOnKotlinChange() throws MavenInvocat
     }
 
+    @Test
+    public void testThatTheApplicationIsReloadedOnScalaChange() throws MavenInvocationException, IOException {
```
By nature, these tests IT take some minutes  to run,  I wonder if this test should be a unit test using `QuarkusDevModeTest` since https://github.com/quarkusio/quarkus/pull/3317 has already landed. 

@gsmet WDYT?

geoand replied:
If `QuarkusDevModeTest` can cover this case, it would be ideal to use it. 
And if it does, then replacing the existing tests with something that uses `QuarkusDevModeTest` would be even better :)

machi1990 replied:
Sure,  I think tests that check if files have been correctly modified, deleted etc can leverage `QuarkusDevModeTest`. We have an issue for that?

geoand replied:
Not that I know of, but I haven't really checked

gsmet commented:
Merged, thanks! Very nice addition.
